lib/segment/src/segment/snapshot.rs

```diff
@@ -47,6 +49,42 @@ impl SnapshotEntry for Segment {
         // flush segment to capture latest state
         self.flush(true, false)?;
 
+        let include_files = match manifest {
+            None => HashSet::new(),
+
+            Some(manifest) => {
+                let updated_manifest = self.get_segment_manifest()?;
+
+                let updated_manifest_json =
+                    serde_json::to_vec(&updated_manifest).map_err(|err| {
+                        OperationError::service_error(format!(
+                            "failed to serialize segment manifest into JSON: {err}"
+                        ))
+                    })?;
+
+                let tar = tar.descend(Path::new(&segment_id))?;
+                tar.blocking_append_data(
+                    &updated_manifest_json,
+                    Path::new("segment_manifest.json"),
+                )?;
+
+                let mut empty_manifest = None;
+                let request_manifest = manifest
+                    .get(segment_id)
+                    .unwrap_or_else(|| empty_manifest.insert(SegmentManifest::empty(segment_id)));
```

Comment:
I don't really get how this works. Shouldn't we take out the `empty_manifest` in between changing it to:

```rust
                let request_manifest = manifest
                    .get(segment_id)
                    .unwrap_or_else(|| SegmentManifest::empty(segment_id));
```

Reply:
`manifest` is `Option<&SegmentManifest>`. We can't return `&SegmentManifest::empty(...)` from `unwrap_or_else`, so we either have to clone `manifest` or use this trick to _conditionally_ initialize empty manifest (into `empty_manifest` variable) and return reference to this empty manifest from `unwrap_or_else` call.

Reply:
Ah, I get it, it's because of the reference.

Construction of `SegmentManifest::empty(segment_id)` is cheap though, it's only a `&str -> String`.